src/main/java/com/atlassian/bitbucket/jenkins/internal/deployments/DeployedToEnvironmentNotifierStep.java

```diff
@@ -0,0 +1,242 @@
+        return environment.getType();
+    }
+
+    @CheckForNull
+    public String getEnvironmentUrl() {
+        return environment.getUrl();
+    }
+
+    @Override
+    public void perform(Run<?, ?> run, FilePath workspace, Launcher launcher,
+                        TaskListener listener) throws InterruptedException, IOException {
+        try {
+            BitbucketRevisionAction revisionAction = run.getAction(BitbucketRevisionAction.class);
+            if (revisionAction == null) {
+                // Not checked out with a Bitbucket SCM
+                listener.error("Could not send deployment notification: DeployedToEnvironmentNotifierStep only works when using the Bitbucket SCM for checkout.");
+                return;
+            }
+
+            // We use the default call to the bitbucketDeploymentFactory to get the state of the deployment based
+            // on the run.
+            BitbucketDeployment deployment =
+                    descriptor().getBitbucketDeploymentFactory().createDeployment(run, environment);
+
+            BitbucketSCMRepository bitbucketSCMRepo = revisionAction.getBitbucketSCMRepo();
+            String revisionSha = revisionAction.getRevisionSha1();
+            descriptor().getDeploymentPoster().postDeployment(bitbucketSCMRepo.getServerId(),
+                    bitbucketSCMRepo.getProjectKey(), bitbucketSCMRepo.getRepositorySlug(), revisionSha, deployment,
+                    run, listener);
+        } catch (RuntimeException e) {
+            // This shouldn't happen because deploymentPoster.postDeployment doesn't throw anything. But just in case,
+            // we don't want to throw anything and potentially stop other steps from being executed
+            String errorMsg =
+                    String.format("An exception occurred when trying to post the deployment to Bitbucket Server: %s", e.getMessage());
+            listener.error(errorMsg);
+            LOGGER.info(errorMsg);
+            LOGGER.log(Level.FINE, "Stacktrace from deployment post failure", e);
+        }
+    }
+
+    private static String getOrGenerateEnvironmentKey(@CheckForNull String environmentKey) {
+        if (!isBlank(environmentKey)) {
+            return environmentKey;
+        }
+        return UUID.randomUUID().toString();
+    }
+
+    private static BitbucketDeploymentEnvironment validateAndGetEnvironment(String environmentKey,
+                                                                            String environmentName,
+                                                                            @CheckForNull String environmentType,
+                                                                            @CheckForNull String environmentUrl) {
+        FormValidation environmentNameValidation = validateEnvironmentName(environmentName);
+        if (!environmentNameValidation.equals(FORM_VALIDATION_OK)) {
+            throw new IllegalArgumentException(environmentNameValidation.getMessage());
```

Comment:
This still suffers from the same problem (but with a much better message), I mentioned in my other comment: https://github.com/jenkinsci/atlassian-bitbucket-server-integration-plugin/pull/265#discussion_r673665818